File: 0153.FindMinimuminRotatedSortedArray.go
func findMin(nums []int) int {
    left, right := 0, len(nums)-1
    for left < right {
        mid := (left+right) >> 1
        if nums[mid] > nums[right] {
            left = mid + 1
        } else {
            right = mid
        }
    }
    /*
    for left < right {
        if nums[left] <= nums[right] {
            break
        }
        mid := (left+right) >> 1
        if nums[mid] >= nums[left] {
            left = mid + 1
        } else {
            right = mid
        }
    }*/
    return nums[left]
}

File: 0000-0999/0153.FindMinimuminRotatedSortedArray.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        left, right = 0, len(nums)-1
        while left < right:
            '''# slow
            mid = (left+right) / 2
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid
            '''
            # fast
            if nums[left] < nums[right]:
                break
            mid = (left+right) / 2
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid
        return nums[left]
